cpp/include/raft/linalg/reduce_rows_by_key.cuh

```diff
@@ -54,21 +56,27 @@ void convert_array(IteratorT1 dst, IteratorT2 src, int n, cudaStream_t st)
  * @param[in]  nkeys       Number of unique keys in d_keys
  * @param[out] d_sums      Row sums by key (ncols x d_keys)
  * @param[in]  stream      CUDA stream
+ * @param[in]  reset_sums  Whether to reset the output sums to zero before reducing
  */
-template <typename DataIteratorT, typename KeysIteratorT, typename WeightT>
-void reduce_rows_by_key(const DataIteratorT* d_A,
-                        int lda,
+template <typename DataIteratorT,
+          typename KeysIteratorT,
+          typename WeightT,
+          typename SumsT,
+          typename IdxT>
+void reduce_rows_by_key(const DataIteratorT d_A,
+                        IdxT lda,
                         const KeysIteratorT d_keys,
                         const WeightT* d_weights,
                         char* d_keys_char,
-                        int nrows,
-                        int ncols,
-                        int nkeys,
-                        DataIteratorT* d_sums,
-                        cudaStream_t stream)
+                        IdxT nrows,
```

Comment:
We are going to be deprecating the raw pointer APIs soon in favor of the new (more self-documenting) mdspan APIs. Do you see any reasons why we should prefer to keep the iterator-based APIs over the mdspan APIs?

Reply:
Yeah, I get the point of the "streamability" of the iterators vs raw pointers. Given that the mdspan is really just a very lightweight wrapper around any type (it can be but doesn't have to be a raw pointer) and we're just forwarding the underlying `data_handle()` to the function in the `detail` namespace, can't we also wrap the mdspan around an iterator?

Reply:
> mdspan is really just a very lightweight wrapper around any type (it can be but doesn't have to be a raw pointer)

Not exactly. As far as I understand, span or mdspan can only wrap around data that exists in memory (and continuously). `std::span` even has a member `data` which returns a pointer to the memory location of the first element. So it can wrap around simple iterators like `vector::begin` but not `cub::TransformInputIterator`.

Reply:
mdspan and span are a little different in both form and function. we have been investing in mdspan more than span for our public APIs. mdspan itself does not require contiguity, though we do enforce that in many of the public APIs for the simple reason that we are using it as a facade for raw pointers. If we are expecting iterators, we could relax that constraint a bit when the underlying data_handle() is in fact an iterator. Also, @mhoemmen can correct me if I'm wrong but I don't believe the data backing an mdspan does need to exist in memory- I think it could even be a file pointer or a pointer to a remote data buffer so long as the accessor knows how to materialize it and it allows for random access. 

Im not trying to discourage the use of iterators in our public API and I definitely see the benefits of using them to apply functors lazily upon materialization to avoid copies / additional allocations. I'm hoping we can incorporate acceptance of iterators more consistently and broadly across our APIs and avoid having a only a couple functions throughout the codebase that accept iterators while most do not.

Reply:
I'm also okay accepting this PR and adding a todo to figure out how we can accept iterator types more broadly across the new APIs so we can consolidate the dev experience a bit for our end-users.

Reply:
Oh, I thought of `mdspan` as a multi-dimensional `std::span` but it seems that it's a bit more versatile.
So, to make sure I understand: each custom iterator would require a different accessor? And the mdspan type for iterator-accepting args will still be a template parameter, while args expecting raw pointers can use matrix or vector views?

Reply:
> Also, @mhoemmen can correct me if I'm wrong but I don't believe the data backing an mdspan does need to exist in memory- I think it could even be a file pointer or a pointer to a remote data buffer so long as the accessor knows how to materialize it and it allows for random access.

@cjnolet is correct (as usual!).  We say "mdspan doesn't need a backing span."

1. Elements don't need to exist in memory;
2. `data_handle_type` doesn't need to be `ElementType*`;
3. even if it is, or even if `data_handle_type` presents the syntax of an iterator, it doesn't need to act like one;
4. `reference` doesn't need to be `element_type&`; and
5. `access(p, i)` doesn't need to return `p[i]`.

One example of (3) is `MPI_Win` (which could be `void*`, but it's not a pointer to an array), a handle to remote memory.

Reply:
Also, it looks like all of RAFT's rank-1 mdspan would let one use `data_handle()`, `data_handle() + extent(0)` as an iterator range.